File: Digitale Signalverarbeitung (Matlab)/Kapitel 3/MATLAB/BEISPIELE/echoKompensation.m
% Echokompensation
% durch einen adaptiver Prädiktor 
% (c) HSM, B. Wir 2.8.2012 
clear all; close all;
%% Simulation der Störung
Z = 2000 %...Werte
x = randn(1,Z); x(end-50) = 8;
y = filter([1 0 0 0.8],1,x);

figure(1);
         subplot(3,9, 1:6);
plot(y(end-99:end),'r'), hold on; plot(x(end-99:end));  hold off; 
title('Zufallssignal: blau, mit Echo: rot'); 
%% Echounterdrückung
% Theorie: G(z)[1 - z^{-Delay)(W(z)] = 1
%
% für      G(z) = (1 + 0.8 z^-3) und Delay = 2
%
%  ==>     W(z) =  0.8 z^-1  - 0.64 z^-4   + 0.5 z^-7  - .....
%
%  ==>     w[k] =  [0  0.8  0  0  -0.64 0 0 0.5 ....]  

%% Adaptiver Prädiktor
% 
N = 10;   % Filterordnung+1
DELAY = 2;
STEP = 0.001;

w = zeros(1,N);
[echo_dach,e,w] = lmsPredictor(w,DELAY, STEP,y);

     subplot(3,9,10:15); 
plot(e(end-99:end),'b'); hold on ; %error == Nutzsignal !! 
plot(x(end-99:end)-e(end-99:end),'g'); hold off; % Rekonstruktionsfehler
title('Echokompensation:blau, Rekonstruktionsfehler:grün');
     subplot(3,9,19:24);
stem(w); title('Filterkoeffizienten');

WINDOW = 64;
     subplot(3,9,[7 16 25]);
spectrogram(x,WINDOW);         title('Spektrogramm/dB')
     subplot(3,9,[8 17 26]); 
spectrogram(y,WINDOW);         title('... mit Echo'); axis off;
     subplot(3,9,[9 18 27]); 
spectrogram(echo_dach,WINDOW); title('... Rek.Fehler'); axis off;

%% Hinweis: Adaptive ARMA-Filter und RLS-Abgleich
%  sind für die Anwendung (noch) besser geeignet. 

File: (Matlab) Digitale Signalverarbeitung/Kapitel 3/Python/BEISPIELE/echoKompensation.py
# -*- coding: utf-8 -*-
"""
 Echokompensation
 durch einen adaptiver Prädiktor 
 (c) HSM, B. Wir 2.8.2012, 
     Jorge Zuniga Le-Bert, Hochschule Mannheim 20.7.2021, (MATLAB to Python)
"""
import numpy as np
from adaptivfilter import lmsPredictor
import scipy.signal as signal               # signal.lfilter()
import matplotlib.pyplot as plt

#%% Simulation der Störung

Z = 4000                                    # Anzahl von Werte
x = np.random.randn(Z)
x[-50] = 8                                  # MATLAB: x(end-50) = 8                             
t = np.arange(0,100,1)                      # 100 k-Werte
y = signal.lfilter([1., 0, 0, 0, 0.8],1,x)
plt.figure(1)
plt.subplot(311)
plt.subplots_adjust(hspace=.8)
plt.plot(t,y[-100:], t, x[-100:],'r')
plt.title('Zufallssignal: rot, mit Echo: blau');
plt.xlabel('K -->')


#%% Echounterdrückung
"""
 Theorie: G(z)[1 - z^{-Delay)(W(z)] = 1

 für      G(z) = (1 + 0.8 z^-3) und Delay = 2

  ==>     W(z) =  0.8 z^-1  - 0.64 z^-4   + 0.5 z^-7  - .....

  ==>     w[k] =  [0  0.8  0  0  -0.64 0 0 0.5 ....]  
"""
#%% Adaptiver Prädiktor
N = 11                     # Filterordnung+1
DELAY = 2
STEP = 0.001    # 0.001
w = np.zeros(N)
echo_dach,e,w = lmsPredictor(w, DELAY, STEP, y)

plt.subplot(312)
plt.subplots_adjust(hspace=1.2)
plt.plot(t,e[-100:],'b', t, x[-100:]-e[-100:],'g')
#plt.text(0,5,'Echokompensation:blau, Rekonstruktionsfehler:grün');
plt.title('Echokompensation:blau, Rekonstruktionsfehler:grün');
plt.xlabel('K -->')
plt.subplot(313)
plt.stem(w)
plt.title('Koeffizienten w')
plt.xlabel('N -->')
print('Koeffizienten w = ',w)




